smithy-typescript-ssdk-libs/server-common/src/validation/validators.ts

```diff
@@ -0,0 +1,153 @@
+        if (min === undefined && max === undefined) {
+            throw new Error('Length constraints must have at least a min or a max.');
+        }
+        this.min = min;
+        this.max = max;
+    }
+
+    validate(input: { length: number } | { [key: string]: any }, memberName: string): LengthValidationFailure | null {
+        let length: number;
+        if (LengthValidator.hasProperty(input, 'length')) {
+            length = input.length;
+        } else {
+            length = Object.keys(input).length;
+        }
+
+        if (this.min !== undefined && length < this.min || this.max !== undefined && length > this.max) {
+            return {
+                constraintType: 'length',
+                constraintValues: this.min === undefined ? [undefined, this.max!]
+                                    : this.max === undefined ? [this.min!, undefined]
+                                    : [this.min!, this.max!],
+                memberName: memberName,
+                failureValue: length
+            }
+        }
+
+        return null;
+    }
+
+    private static hasProperty<P extends PropertyKey>(obj: any, prop: P): obj is Record<P, unknown> {
+        return obj.hasOwnProperty(prop)
+    }
+}
+
+export class PatternValidator {
+    private readonly inputPattern: string;
+    private readonly pattern: RegExp;
+
+    constructor(pattern: string) {
+        this.inputPattern = pattern;
+        this.pattern = new RegExp(pattern, "u");
+    }
+
+    validate(input: string, memberName: string): PatternValidationFailure | null {
+        if (!input.match(this.pattern)) {
+            return {
+                constraintType: 'pattern',
+                constraintValues: this.inputPattern,
+                failureValue: input,
+                memberName: memberName
+            }
+        }
+        return null;
+    }
+}
+
+export class RangeValidator {
+    private readonly min?: number;
+    private readonly max?: number;
+
+    constructor(min?: number, max?:number) {
+        if (min === undefined && max === undefined) {
+            throw new Error('Range constraints must have at least a min or a max.');
+        }
+        this.min = min;
+        this.max = max;
+    }
+
+    validate(input: number, memberName: string) {
+        if (this.min !== undefined && input < this.min || this.max !== undefined && input > this.max) {
+            return {
+                constraintType: 'range',
+                constraintValues: this.min === undefined ? [undefined, this.max!]
+                                    : this.max === undefined ? [this.min!, undefined]
+                                    : [this.min!, this.max!],
+                memberName: memberName,
+                failureValue: input
+            }
+        }
+
+        return null;
+    }
+}
+
+export class UniqueItemsValidator {
+
+    validate(input: Array<any>, memberName: string): UniqueItemsValidationFailure | null {
```

Comment:
Not to sound too flippant, but do we care? All the validator is checking for is uniqueness. The type of the items being checked has nothing to do with the validation, does it?

Reply:
Yeah it's ok I guess. Was just trying to understand if it was difficult to codify for same type v/s you didn't do it as it isn't necessary. I thought one generally tried to avoid `any`. Maybe it is appropriate here. Still learning TS best practices.